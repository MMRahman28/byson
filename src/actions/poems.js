import { v4 as uuid } from 'uuid';
import database from '../firebase/firebase';
// ADD_POEM
export const addPoem = (poem) => (
{
    type: 'ADD_POEM',
    poem

  });

export const startAddPoem = (poemData = {}) => {
  return (dispatch) => {
    const {
      note = '',
      createdAt = 0
    } = poemData;

    const poem = {note, createdAt};
  return   database.ref('poems').push(poem).then((ref) => {
      dispatch(addPoem({
        id: ref.key,
        ...poem
      }));
    });
  };
};

// REMOVE_POEM

export const removePoem = ({id} = {}) => (
  {
      type: 'REMOVE_POEM',
      id
  }

);

// EDIT_POEM
export const editPoem = (id, updates) => (
    {
        type: 'EDIT_POEM',
        id,
        updates

    }
);

// SET_POEM
export const setPoem = (poems) => ({
  type: 'SET_POEM',
  poems
});
