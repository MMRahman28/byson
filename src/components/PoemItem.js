// Export a stateless functional component
// description, amount, createdAt
import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import 'numeral/locales/en-gb';



//import { connect } from 'react-redux';
//import {removeExpense} from '../actions/expenses';
//numeral.locale('en-gb');
const PoemItem = ({id, note, createdAt}) => (
 <div>
 <Link to={`/edit/${id}`}>
 <h3>Edit</h3>
 </Link>
   <p>{note} - {moment(createdAt).format('Do MMMM, YYYY')}</p>

  </div>
);


//export default connect()(ExpenseListItem);
export default PoemItem;
