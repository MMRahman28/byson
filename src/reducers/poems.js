import uuid from 'uuid';
//import { v4 as uuid } from 'uuid';
const poemsReducerDefaultState = [];
const randomPoems = [...Array(4)].map(() => Math.floor(Math.random() * 9));
export default (state = poemsReducerDefaultState, action) => {
    switch (action.type)
    {
        case 'ADD_POEM':
            return [...state, action.poem];

        case 'REMOVE_POEM':
            return [...state].filter((poem) => (poem.id !== action.id));
        case 'EDIT_POEM':
            return state.map((poem) => {
                if(poem.id === action.id){

                     return {
                         ...poem,
                         ...action.updates
                     }
                } else {
                    return poem;
                }
            });

        case 'SET_POEMS':
          return action.poems;
          
        // case 'SET_POEM':
        //      return [...state].filter((poem) => (randomPoems.includes(poem.id)));


        default:
            return state;
    }
};
