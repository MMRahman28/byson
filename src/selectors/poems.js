import moment from 'moment';

// expenses: state.expenses
// {text, sortBy, startDate, endate} : state.filters
export default (poems, { note, sortBy, startDate, endDate}) => {
    return poems.filter((poem) => {
         const createdAtMoment = moment(poem.createdAt);
         const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day'): true;


        // figure out if expenses.description has the text variable string inside of it
        // includes
        // convert both strings to lowercase
        const textMatch = typeof note !== 'string' || poem.note.toLowerCase().includes(note.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;

    }).sort((a, b) => {
    //   // sortBy -> date
    //   // put the ones with the latest date first
    if(sortBy === 'date') {
             return a.createdAt < b.createdAt ? 1 : -1;
           }
        });
    //     // sortBy -> amount
    //     // put the ones with a greater amount first
    //     else if(sortBy === 'amount') {
    //         return a.amount < b.amount ? 1 : -1;
    //     }
    // });

};
