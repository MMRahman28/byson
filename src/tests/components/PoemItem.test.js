import React from 'react';
import { shallow } from 'enzyme';
import poems from '../fixtures/poems';
import PoemItem from '../../components/PoemItem';

test('should render PoemItem correctly', () => {
  const wrapper = shallow(<PoemItem {...poems[0]} />);
  expect(wrapper).toMatchSnapshot();
});
