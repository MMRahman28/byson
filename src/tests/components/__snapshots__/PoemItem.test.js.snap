// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PoemItem correctly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Edit
    </h3>
  </Link>
  <p>
    Morning Poem
     - 
    1st January, 1970
  </p>
</div>
`;
